### ISSUE TITLE
[BUG] error while getting associated traintuples to update their inModel

### ISSUE THREAD

**Issue by jmorel** [Substra/substra-chaincode#65]
**Describe the bug**

While creating a fairly large compute plan, my first tuple of the plan is stuck in doing while all others are stuck in waiting.

**Logs**  

Here are the chaincode logs after the creation of the compute plan.
```
2019-12-16 15:55:09.926 UTC [substra-chaincode] Infof -> INFO 02f Args received by the chaincode: []string{"logStartCompositeTrain", "{\"key\": \"a43a2efc1415b16618d1a1349eb9c89b5d6bee0849a3395bc9af3504cebad19d\"}"}
2019-12-16 15:55:09.931 UTC [substra-chaincode] Infof -> INFO 030 traintuple a43a2efc1415b16618d1a1349eb9c89b5d6bee0849a3395bc9af3504cebad19d status updated: doing (from=todo)
2019-12-16 15:55:09.931 UTC [substra-chaincode] Infof -> INFO 031 Response from chaincode: main.outputCompositeTraintuple{Key:"a43a2efc1415b16618d1a1349eb9c89b5d6bee0849a3395bc9af3504cebad19d", Algo:(*main.HashDressName)(0xc0001b8300), Creator:"MyOrg1MSP", Dataset:(*main.outputTtDataset)(0xc00033b140), ComputePlanID:"a43a2efc1415b16618d1a1349eb9c89b5d6bee0849a3395bc9af3504cebad19d", InHeadModel:(*main.Model)(nil), InTrunkModel:(*main.Model)(nil), Log:"", OutHeadModel:main.outModelComposite{OutModel:(*main.HashDress)(nil), Permissions:main.outputPermissions{Process:main.Permission{Public:false, AuthorizedIDs:[]string{"MyOrg1MSP"}}}}, OutTrunkModel:main.outModelComposite{OutModel:(*main.HashDress)(nil), Permissions:main.outputPermissions{Process:main.Permission{Public:false, AuthorizedIDs:[]string{"MyOrg1MSP"}}}}, Rank:0, Status:"doing", Tag:""}, error: %!s(<nil>)
2019-12-16 15:55:17.787 UTC [substra-chaincode] Infof -> INFO 032 Args received by the chaincode: []string{"logSuccessCompositeTrain", "{\"key\": \"a43a2efc1415b16618d1a1349eb9c89b5d6bee0849a3395bc9af3504cebad19d\", \"log\": \"\", \"outHeadModel\": {\"hash\": \"859689a853f540aa56a11fb6201bf217d6cebdbe4fa3f605f4b4dc020d4f606f\", \"storageAddress\": \"http://substra-backend.node-1.com/model/859689a853f540aa56a11fb6201bf217d6cebdbe4fa3f605f4b4dc020d4f606f/file/\"}, \"outTrunkModel\": {\"hash\": \"7da2fbe3bd8a9e1336f919d78c723cfb489a665613913e27b50d454a72831af2\", \"storageAddress\": \"http://substra-backend.node-1.com/model/7da2fbe3bd8a9e1336f919d78c723cfb489a665613913e27b50d454a72831af2/file/\"}}"}
2019-12-16 15:55:17.805 UTC [substra-chaincode] Infof -> INFO 033 traintuple a43a2efc1415b16618d1a1349eb9c89b5d6bee0849a3395bc9af3504cebad19d status updated: done (from=doing)
2019-12-16 15:55:17.815 UTC [substra-chaincode] Infof -> INFO 034 traintuple d60e73cc04e9ab853da162ae70eb7cbae90969746c9de350f7084a10087788d5 status updated: todo (from=waiting)
2019-12-16 15:55:47.797 UTC [shim] handleQueryStateClose -> ERRO 035 [f4aaed5c] Received ERROR
2019-12-16 15:55:47.809 UTC [shim] handleGetStateByRange -> ERRO 036 [f4aaed5c] Received ERROR
2019-12-16 15:55:47.809 UTC [substra-chaincode] Infof -> INFO 037 Response from chaincode: main.outputCompositeTraintuple{Key:"", Algo:(*main.HashDressName)(nil), Creator:"", Dataset:(*main.outputTtDataset)(nil), ComputePlanID:"", InHeadModel:(*main.Model)(nil), InTrunkModel:(*main.Model)(nil), Log:"", OutHeadModel:main.outModelComposite{OutModel:(*main.HashDress)(nil), Permissions:main.outputPermissions{Process:main.Permission{Public:false, AuthorizedIDs:[]string(nil)}}}, OutTrunkModel:main.outModelComposite{OutModel:(*main.HashDress)(nil), Permissions:main.outputPermissions{Process:main.Permission{Public:false, AuthorizedIDs:[]string(nil)}}}, Rank:0, Status:"", Tag:""}, error: error while getting associated traintuples to update their inModel
2019-12-16 15:55:49.845 UTC [substra-chaincode] Infof -> INFO 038 Args received by the chaincode: []string{"logSuccessCompositeTrain", "{\"key\": \"a43a2efc1415b16618d1a1349eb9c89b5d6bee0849a3395bc9af3504cebad19d\", \"log\": \"\", \"outHeadModel\": {\"hash\": \"859689a853f540aa56a11fb6201bf217d6cebdbe4fa3f605f4b4dc020d4f606f\", \"storageAddress\": \"http://substra-backend.node-1.com/model/859689a853f540aa56a11fb6201bf217d6cebdbe4fa3f605f4b4dc020d4f606f/file/\"}, \"outTrunkModel\": {\"hash\": \"7da2fbe3bd8a9e1336f919d78c723cfb489a665613913e27b50d454a72831af2\", \"storageAddress\": \"http://substra-backend.node-1.com/model/7da2fbe3bd8a9e1336f919d78c723cfb489a665613913e27b50d454a72831af2/file/\"}}"}
2019-12-16 15:55:49.852 UTC [substra-chaincode] Infof -> INFO 039 traintuple a43a2efc1415b16618d1a1349eb9c89b5d6bee0849a3395bc9af3504cebad19d status updated: done (from=doing)
2019-12-16 15:55:49.861 UTC [substra-chaincode] Infof -> INFO 03a traintuple d60e73cc04e9ab853da162ae70eb7cbae90969746c9de350f7084a10087788d5 status updated: todo (from=waiting)
2019-12-16 15:56:19.813 UTC [shim] handleQueryStateClose -> ERRO 03b [00b2f41a] Received ERROR
2019-12-16 15:56:19.814 UTC [shim] handleGetStateByRange -> ERRO 03c [00b2f41a] Received ERROR
2019-12-16 15:56:19.814 UTC [substra-chaincode] Infof -> INFO 03d Response from chaincode: main.outputCompositeTraintuple{Key:"", Algo:(*main.HashDressName)(nil), Creator:"", Dataset:(*main.outputTtDataset)(nil), ComputePlanID:"", InHeadModel:(*main.Model)(nil), InTrunkModel:(*main.Model)(nil), Log:"", OutHeadModel:main.outModelComposite{OutModel:(*main.HashDress)(nil), Permissions:main.outputPermissions{Process:main.Permission{Public:false, AuthorizedIDs:[]string(nil)}}}, OutTrunkModel:main.outModelComposite{OutModel:(*main.HashDress)(nil), Permissions:main.outputPermissions{Process:main.Permission{Public:false, AuthorizedIDs:[]string(nil)}}}, Rank:0, Status:"", Tag:""}, error: error while getting associated composite traintuples to update their inModel
2019-12-16 15:56:23.886 UTC [substra-chaincode] Infof -> INFO 03e Args received by the chaincode: []string{"logSuccessCompositeTrain", "{\"key\": \"a43a2efc1415b16618d1a1349eb9c89b5d6bee0849a3395bc9af3504cebad19d\", \"log\": \"\", \"outHeadModel\": {\"hash\": \"859689a853f540aa56a11fb6201bf217d6cebdbe4fa3f605f4b4dc020d4f606f\", \"storageAddress\": \"http://substra-backend.node-1.com/model/859689a853f540aa56a11fb6201bf217d6cebdbe4fa3f605f4b4dc020d4f606f/file/\"}, \"outTrunkModel\": {\"hash\": \"7da2fbe3bd8a9e1336f919d78c723cfb489a665613913e27b50d454a72831af2\", \"storageAddress\": \"http://substra-backend.node-1.com/model/7da2fbe3bd8a9e1336f919d78c723cfb489a665613913e27b50d454a72831af2/file/\"}}"}
2019-12-16 15:56:23.890 UTC [substra-chaincode] Infof -> INFO 03f traintuple a43a2efc1415b16618d1a1349eb9c89b5d6bee0849a3395bc9af3504cebad19d status updated: done (from=doing)
2019-12-16 15:56:23.898 UTC [substra-chaincode] Infof -> INFO 040 traintuple d60e73cc04e9ab853da162ae70eb7cbae90969746c9de350f7084a10087788d5 status updated: todo (from=waiting)
2019-12-16 15:56:53.853 UTC [shim] handleGetStateByRange -> ERRO 041 [8fca8577] Received ERROR
2019-12-16 15:56:53.853 UTC [substra-chaincode] Infof -> INFO 042 Response from chaincode: main.outputCompositeTraintuple{Key:"", Algo:(*main.HashDressName)(nil), Creator:"", Dataset:(*main.outputTtDataset)(nil), ComputePlanID:"", InHeadModel:(*main.Model)(nil), InTrunkModel:(*main.Model)(nil), Log:"", OutHeadModel:main.outModelComposite{OutModel:(*main.HashDress)(nil), Permissions:main.outputPermissions{Process:main.Permission{Public:false, AuthorizedIDs:[]string(nil)}}}, OutTrunkModel:main.outModelComposite{OutModel:(*main.HashDress)(nil), Permissions:main.outputPermissions{Process:main.Permission{Public:false, AuthorizedIDs:[]string(nil)}}}, Rank:0, Status:"", Tag:""}, error: error while getting associated aggregate tuples to update their inModel
2019-12-16 15:57:01.931 UTC [substra-chaincode] Infof -> INFO 043 Args received by the chaincode: []string{"logSuccessCompositeTrain", "{\"key\": \"a43a2efc1415b16618d1a1349eb9c89b5d6bee0849a3395bc9af3504cebad19d\", \"log\": \"\", \"outHeadModel\": {\"hash\": \"859689a853f540aa56a11fb6201bf217d6cebdbe4fa3f605f4b4dc020d4f606f\", \"storageAddress\": \"http://substra-backend.node-1.com/model/859689a853f540aa56a11fb6201bf217d6cebdbe4fa3f605f4b4dc020d4f606f/file/\"}, \"outTrunkModel\": {\"hash\": \"7da2fbe3bd8a9e1336f919d78c723cfb489a665613913e27b50d454a72831af2\", \"storageAddress\": \"http://substra-backend.node-1.com/model/7da2fbe3bd8a9e1336f919d78c723cfb489a665613913e27b50d454a72831af2/file/\"}}"}
2019-12-16 15:57:01.934 UTC [substra-chaincode] Infof -> INFO 044 traintuple a43a2efc1415b16618d1a1349eb9c89b5d6bee0849a3395bc9af3504cebad19d status updated: done (from=doing)
2019-12-16 15:57:01.946 UTC [substra-chaincode] Infof -> INFO 045 traintuple d60e73cc04e9ab853da162ae70eb7cbae90969746c9de350f7084a10087788d5 status updated: todo (from=waiting)





2019-12-16 15:57:31.898 UTC [shim] handleQueryStateClose -> ERRO 046 [ce4dbc58] Received ERROR
2019-12-16 15:57:31.900 UTC [shim] handleGetStateByRange -> ERRO 047 [ce4dbc58] Received ERROR
2019-12-16 15:57:31.901 UTC [substra-chaincode] Infof -> INFO 048 Response from chaincode: main.outputCompositeTraintuple{Key:"", Algo:(*main.HashDressName)(nil), Creator:"", Dataset:(*main.outputTtDataset)(nil), ComputePlanID:"", InHeadModel:(*main.Model)(nil), InTrunkModel:(*main.Model)(nil), Log:"", OutHeadModel:main.outModelComposite{OutModel:(*main.HashDress)(nil), Permissions:main.outputPermissions{Process:main.Permission{Public:false, AuthorizedIDs:[]string(nil)}}}, OutTrunkModel:main.outModelComposite{OutModel:(*main.HashDress)(nil), Permissions:main.outputPermissions{Process:main.Permission{Public:false, AuthorizedIDs:[]string(nil)}}}, Rank:0, Status:"", Tag:""}, error: error while getting associated aggregate tuples to update their inModel
2019-12-16 15:57:47.935 UTC [substra-chaincode] Infof -> INFO 049 Args received by the chaincode: []string{"logSuccessCompositeTrain", "{\"key\": \"a43a2efc1415b16618d1a1349eb9c89b5d6bee0849a3395bc9af3504cebad19d\", \"log\": \"\", \"outHeadModel\": {\"hash\": \"859689a853f540aa56a11fb6201bf217d6cebdbe4fa3f605f4b4dc020d4f606f\", \"storageAddress\": \"http://substra-backend.node-1.com/model/859689a853f540aa56a11fb6201bf217d6cebdbe4fa3f605f4b4dc020d4f606f/file/\"}, \"outTrunkModel\": {\"hash\": \"7da2fbe3bd8a9e1336f919d78c723cfb489a665613913e27b50d454a72831af2\", \"storageAddress\": \"http://substra-backend.node-1.com/model/7da2fbe3bd8a9e1336f919d78c723cfb489a665613913e27b50d454a72831af2/file/\"}}"}
2019-12-16 15:57:47.939 UTC [substra-chaincode] Infof -> INFO 04a traintuple a43a2efc1415b16618d1a1349eb9c89b5d6bee0849a3395bc9af3504cebad19d status updated: done (from=doing)
2019-12-16 15:57:47.959 UTC [substra-chaincode] Infof -> INFO 04b traintuple d60e73cc04e9ab853da162ae70eb7cbae90969746c9de350f7084a10087788d5 status updated: todo (from=waiting)
2019-12-16 15:58:17.904 UTC [shim] handleQueryStateClose -> ERRO 04c [a21cd41b] Received ERROR
2019-12-16 15:58:17.905 UTC [shim] handleGetStateByRange -> ERRO 04d [a21cd41b] Received ERROR
2019-12-16 15:58:17.905 UTC [substra-chaincode] Infof -> INFO 04e Response from chaincode: main.outputCompositeTraintuple{Key:"", Algo:(*main.HashDressName)(nil), Creator:"", Dataset:(*main.outputTtDataset)(nil), ComputePlanID:"", InHeadModel:(*main.Model)(nil), InTrunkModel:(*main.Model)(nil), Log:"", OutHeadModel:main.outModelComposite{OutModel:(*main.HashDress)(nil), Permissions:main.outputPermissions{Process:main.Permission{Public:false, AuthorizedIDs:[]string(nil)}}}, OutTrunkModel:main.outModelComposite{OutModel:(*main.HashDress)(nil), Permissions:main.outputPermissions{Process:main.Permission{Public:false, AuthorizedIDs:[]string(nil)}}}, Rank:0, Status:"", Tag:""}, error: error while getting associated composite traintuples to update their inModel
2019-12-16 15:58:49.964 UTC [substra-chaincode] Infof -> INFO 04f Args received by the chaincode: []string{"logSuccessCompositeTrain", "{\"key\": \"a43a2efc1415b16618d1a1349eb9c89b5d6bee0849a3395bc9af3504cebad19d\", \"log\": \"\", \"outHeadModel\": {\"hash\": \"859689a853f540aa56a11fb6201bf217d6cebdbe4fa3f605f4b4dc020d4f606f\", \"storageAddress\": \"http://substra-backend.node-1.com/model/859689a853f540aa56a11fb6201bf217d6cebdbe4fa3f605f4b4dc020d4f606f/file/\"}, \"outTrunkModel\": {\"hash\": \"7da2fbe3bd8a9e1336f919d78c723cfb489a665613913e27b50d454a72831af2\", \"storageAddress\": \"http://substra-backend.node-1.com/model/7da2fbe3bd8a9e1336f919d78c723cfb489a665613913e27b50d454a72831af2/file/\"}}"}
2019-12-16 15:58:49.969 UTC [substra-chaincode] Infof -> INFO 050 traintuple a43a2efc1415b16618d1a1349eb9c89b5d6bee0849a3395bc9af3504cebad19d status updated: done (from=doing)
2019-12-16 15:58:49.980 UTC [substra-chaincode] Infof -> INFO 051 traintuple d60e73cc04e9ab853da162ae70eb7cbae90969746c9de350f7084a10087788d5 status updated: todo (from=waiting)
2019-12-16 15:59:19.936 UTC [shim] handleQueryStateClose -> ERRO 052 [ca245ba6] Received ERROR
2019-12-16 15:59:19.937 UTC [shim] handleGetStateByRange -> ERRO 053 [ca245ba6] Received ERROR
2019-12-16 15:59:19.937 UTC [substra-chaincode] Infof -> INFO 054 Response from chaincode: main.outputCompositeTraintuple{Key:"", Algo:(*main.HashDressName)(nil), Creator:"", Dataset:(*main.outputTtDataset)(nil), ComputePlanID:"", InHeadModel:(*main.Model)(nil), InTrunkModel:(*main.Model)(nil), Log:"", OutHeadModel:main.outModelComposite{OutModel:(*main.HashDress)(nil), Permissions:main.outputPermissions{Process:main.Permission{Public:false, AuthorizedIDs:[]string(nil)}}}, OutTrunkModel:main.outModelComposite{OutModel:(*main.HashDress)(nil), Permissions:main.outputPermissions{Process:main.Permission{Public:false, AuthorizedIDs:[]string(nil)}}}, Rank:0, Status:"", Tag:""}, error: error while getting associated traintuples to update their inModel
```

In the worker's logs, I repeateadly have the following:
```
[2019-12-16 16:03:01,907: WARNING/ForkPoolWorker-1] Function update_ledger failed (<class 'substrapp.ledger_utils.LedgerBadResponse'>): failed to execute transaction 8a5a13ab100b92f7d74b239ae9bca1f1b3fe6285699b557343ec327a825e83ee: error sending: timeout expired while executing transaction retrying in 128s
```

**Environment:**
- chaincode master / 9a1a0e3
- backend master / 3fce3a27

**Comment by jmorel** [comment]
I reproduced the error on another compute plan and it may be linked to permissions issues.

The following shows 2 chained composite traintuples, the first one executed on Org1 and the second one or Org2. However the first one is private (can only be processed by Org1). The execution of the second one should therefore fail, but what happens is that the first one never changes status from doing to done because of the bug reported above.

```shell
$ substra get composite_traintuple 3cdc6b8b93bb71507ae6b768c6b563d2d36a8ee28c5e4d76b55eb668d96e988c
KEY                             3cdc6b8b93bb71507ae6b768c6b563d2d36a8ee28c5e4d76b55eb668d96e988c
OUT HEAD MODEL KEY              None
OUT HEAD MODEL PERMISSIONS      Processable by:
                                - MyOrg1MSP
OUT TRUNK MODEL KEY             None
OUT TRUNK MODEL PERMISSIONS     Processable by:
                                - MyOrg1MSP
COMPOSITE ALGO KEY              26042d04c4b874e3e626c3e79596767dc8f06084da15639b35aebb209e21ef7b
COMPOSITE ALGO NAME             test_melloddy - Algo 0
STATUS                          doing
DATASET KEY                     71cf692cd0f46bfca9ebafb6835a642f551c09df34ff085c1b50f2fbac26a078
TRAIN DATA SAMPLE KEYS          1 key
IN HEAD MODEL KEY               None
IN TRUNK MODEL KEY              None
RANK                            0
COMPUTE PLAN ID                 3cdc6b8b93bb71507ae6b768c6b563d2d36a8ee28c5e4d76b55eb668d96e988c
TAG
LOG
CREATOR                         MyOrg1MSP
WORKER                          MyOrg1MSP
$ substra get composite_traintuple 58ba3b0119c1e79367cebe77fb707f1a3d9dea0ebca0ab87f8df2a1ef61a0691
KEY                             58ba3b0119c1e79367cebe77fb707f1a3d9dea0ebca0ab87f8df2a1ef61a0691
OUT HEAD MODEL KEY              None
OUT HEAD MODEL PERMISSIONS      Processable by:
                                - MyOrg1MSP
OUT TRUNK MODEL KEY             None
OUT TRUNK MODEL PERMISSIONS     Processable by:
                                - MyOrg1MSP
COMPOSITE ALGO KEY              26042d04c4b874e3e626c3e79596767dc8f06084da15639b35aebb209e21ef7b
COMPOSITE ALGO NAME             test_melloddy - Algo 0
STATUS                          waiting
DATASET KEY                     018897d93d9506ac6f5cf55a0be5e7d54d26c4ef1316102433ee7b139ce7333d
TRAIN DATA SAMPLE KEYS          1 key
IN HEAD MODEL KEY               3cdc6b8b93bb71507ae6b768c6b563d2d36a8ee28c5e4d76b55eb668d96e988c
IN TRUNK MODEL KEY              3cdc6b8b93bb71507ae6b768c6b563d2d36a8ee28c5e4d76b55eb668d96e988c
RANK                            1
COMPUTE PLAN ID                 3cdc6b8b93bb71507ae6b768c6b563d2d36a8ee28c5e4d76b55eb668d96e988c
TAG
LOG
CREATOR                         MyOrg1MSP
WORKER                          MyOrg2MSP
```

**Comment by AurelienGasser** [comment]
@jmorel Could you please provide the script / test you're using to repro this bug?

**Comment by thibaultrobert** [comment]
So far I've not found why we timeout while iterating over a composite key. I've open a PR https://github.com/SubstraFoundation/substra-chaincode/pull/67 with a bit more error logged and a reduction in the number of indexes but it didn't help much on this bug.

**Comment by jmorel** [comment]
In order to reproduce the bug, you can checkout the `load_testing` branch on substra-tests  (https://github.com/SubstraFoundation/substra-tests/pull/50), modify the `COMPUTE_PLAN_SIZES` variable as `COMPUTE_PLAN_SIZES = [8, 9, 10]` and run `pytest tests/test_load.py -rs -v -k test_load_multi_node_composite_aggregates`

The test at 8 iterations (24 tuples) runs fine on my machine but I run in the timeout issue with the 9 iteration test.

I tried also adding a 10s sleep to all algo, it doesn't change anything at all.

**Comment by Kelvin-M** [comment]
Fixed by #69
